e2e-testing.md

```diff
@@ -0,0 +1,321 @@
+  });
+}
+
+
+// ✅ good
+async typeCourseName(name: string): Promise<void> {
+  await test.step('Type course name', async () => {
+    await this.courseDropwDown.type(name);
+  });
+}
+
+```
+
+
+3\. Test scenarios
+--------------
+
+#### 3.1. Use snake_case for test folders naming
+
+
+Use the snake_case for test folder names. Never use spaces in the folders and file names.
+
+
+```typescript
+// ❌ bad
+- tests
+    - e2e
+        - Admin Tools
+            - Homework-Review-Plugin
+
+
+// ✅ good
+- tests
+   - e2e
+        - Admin_Tools
+            - Homework_Review_Plugin
+
+```
+
+
+#### 3.2. Use folders for test suite organization
+
+
+Do not add test spec files from different suites to one folder. Create separate folder for each test suite.
+
+
+#### 3.3. Use one spec file per one test
+
+
+Do not add several tests into one spec file. Each spec file should normally have one test scenario. The exception can be done for the parametrized tests.
+
+
+#### 3.4. Before block should not contain common test data preparation
+
+The before test block in the test spec file should generally contain only calling the fixtures and defining the shortcut constants for better test readability.
+In case one need to define some additional data preparation (for example seeding some data) - one need to do this in a separate methods and fixtures which can be re-used in other tests.
+
+
+
+```typescript
+// ❌ bad
+test.beforeEach((
+  {
+    page,
+    newProfessionUA,
+    techCheckTopicInNewCourse,
+  },
+) => {
+  const techCheckQuestion
+    = new TechCheckQuestion(
+    seedAPI,
+    { techCheckTopic: techCheckTopicInNewCourse },
+  );
+
+  await techCheckQuestion.createWithEnTranslate();
+
+  courseName = newProfessionUA.postpaid.nameShortCode;
+  questionEn = techCheckQuestion.enTranslation;
+
+  questionEditorPage = new QuestionsEditorPage(page);
+});
+
+
+// ✅ good
+test.beforeEach((
+  {
+    page,
+    newProfessionUA,
+    techCheckQuestionInTopicInNewCourse,
+  },
+) => {
+  courseName = newProfessionUA.postpaid.nameShortCode;
+  questionEn = techCheckQuestionInTopicInNewCourse.enTranslation;
+
+  questionEditorPage = new QuestionsEditorPage(page);
+});
+
+```
+
+#### 3.5. Test block should only contain test steps and assertions
+
+Test block should contain only test-step or test-assertion methods - these are page object methods wrapped with allure steps and clearly describing user performed step or assertion.
+Do not use page locators directly in the test spec file - add page-object method instead.
+
+```typescript
+// ❌ bad
+test('should allow to successfully sign in',
+  async () => {
+    await signInPage.visit();
+    await signInPage.emailField.fill(email);
+    ...
+    await signInPage.waitForSelector(successMessage);
+  });
+
+
+// ✅ good
+test('should allow to successfully sign in',
+  async () => {
+    await signInPage.visit();
+    await signInPage.fillUserEmail(email);
+    ...
+    await coursesListPage.assertOpened();
```

Comment:
not clear wha;ts the difference

- visit: in both cases, OK
- emailField.fill vs fillUserEmal - isn't it a matter of taste? or does fillUserEmail contain some other methods like defining allure steps
- waitForSelector vs assertOpened - this is relatively clear, we assert that user is redirected instead of just waiting for selector